src/gtest/test_circuit.cpp

```diff
@@ -74,3 +77,108 @@ TEST(circuit, endianness)
+    wit1.append(commitment2_data);
+    auto wit2 = tree.witness();
+    auto expected_root = tree.root();
+    tree.append(uint256S("3e243c8798678570bb8d42616c23a536af44be15c4eef073490c2a44ae5f32c3"));
+    auto unexpected_root = tree.root();
+    tree.append(uint256S("26d9b20c7f1c3d2528bbcd43cd63344b0afd3b6a0a8ebd37ec51cba34907bec7"));
+    auto badwit1 = tree.witness();
+    tree.append(uint256S("02c2467c9cd15e0d150f74cd636505ed675b0b71b66a719f6f52fdb49a5937bb"));
+    auto badwit2 = tree.witness();
+
+    // Perform the test
+
+    pb.val(commitment1_read) = enforce_a ? FieldT::one() : FieldT::zero();
+    pb.val(commitment2_read) = enforce_b ? FieldT::one() : FieldT::zero();
+
+    commitment1.bits.fill_with_bits(pb, uint256_to_bool_vector(commitment1_data));
+    commitment2.bits.fill_with_bits(pb, uint256_to_bool_vector(commitment2_data));
+
+    if (write_root_first) {
+        root.bits.fill_with_bits(pb, uint256_to_bool_vector(expected_root));
+    }
+
+    mgadget1.generate_r1cs_witness(wit1.path());
+    mgadget2.generate_r1cs_witness(wit2.path());
+
+    // Overwrite with our expected root
+    root.bits.fill_with_bits(pb, uint256_to_bool_vector(expected_root));
+
+    return pb.is_satisfied();
+}
+
+TEST(circuit, merkle_tree_gadget_weirdness)
+{
+    /*
+    The merkle tree gadget takes a leaf in the merkle tree (the Note commitment),
+    a merkle tree authentication path, and a root (anchor). It also takes a parameter
+    called read_success, which is used to determine if the commitment actually needs to
+    appear in the tree.
+
+    If two input notes use the same root (which our protocol does) then if `read_success`
+    is disabled on the first note but enabled on the second note (i.e., the first note
+    has value of zero and second note has nonzero value) then there is an edge case in
+    the witnessing behavior. The first witness will accidentally constrain the root to
+    equal null (the default value of the anchor) and the second witness will actually
+    copy the bits, violating the constraint system.
+
+    Notice that this edge case is not in the constraint system but in the witnessing
+    behavior.
+    */
+
+    default_r1cs_ppzksnark_pp::init_public_params();
+    typedef Fr<default_r1cs_ppzksnark_pp> FieldT;
+
+    // Test the normal case
+    ASSERT_TRUE(test_merkle_gadget<FieldT>(true, true, false));
+    ASSERT_TRUE(test_merkle_gadget<FieldT>(true, true, true));
+
+    // Test the case where the first commitment is enforced but the second isn't
+    // Works because the first read is performed before the second one
+    ASSERT_TRUE(test_merkle_gadget<FieldT>(true, false, false));
+    ASSERT_TRUE(test_merkle_gadget<FieldT>(true, false, true));
+
+    // Test the case where the first commitment isn't enforced but the second is
+    // Doesn't work because the first multipacker witnesses the existing root (which
+    // is null)
+    ASSERT_TRUE(!test_merkle_gadget<FieldT>(false, true, false));
```

Comment:
Is there an ASSERT_FALSE? (I was about to write "Should this be ASSERT_FALSE?" then I noticed the `!`).